### PR TITLE
prov/efa: fix a performance issue caused by lazy initialization of packet pool

### PR DESCRIPTION
**Pull request by wzamazon**
This PR contains two commits that fix a performance issue caused by lazy initialization of packet pool

### CODE REVIEW

File: prov/efa/src/rxr/rxr_ep.c [hunk 1]
```diff
@@ -1090,6 +1098,14 @@ int rxr_ep_init(struct rxr_ep *ep)
 
 		if (ret)
 			goto err_free;
+
+		ret = ofi_bufpool_grow(ep->rx_ooo_pkt_pool);
```
Comment (rajachan) [inline prov/efa/src/rxr/rxr_ep.c:1102]: If it does not already, `ofi_bufpool_create` should allow the caller to decide if they want lazy or immediate allocation.

**Review by rajachan — APPROVED** [review]

**Review by rajachan — COMMENTED** [review]
Accepted in error.

File: prov/efa/src/rxr/rxr_ep.c (hunk 1 above)
Comment (wckzhang) [reply prov/efa/src/rxr/rxr_ep.c:1102]: As far as I can tell, the function does not allow this option. Are you asking for a flag and check to be added in the bufpool create function to do so?

Comment (rajachan) [reply prov/efa/src/rxr/rxr_ep.c:1102]: Yeah, pretty much. This could be an attribute flag for the bufpool.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:1104]: @rajachan Such a change would require changing prototype of `ofi_bufpool_create`, which requires changing multiple provider (rxm, psm3). So I prefer to do that after 1.13.0 release.

**Comment by wzamazon** [comment]
Added a commit to adjust ooo pool chunk size.

File: prov/efa/src/rxr/rxr_ep.c [hunk 2]
```diff
@@ -1093,8 +1093,10 @@ int rxr_ep_init(struct rxr_ep *ep)
 
 	if (rxr_env.rx_copy_ooo) {
 		ret = ofi_bufpool_create(&ep->rx_ooo_pkt_pool, entry_sz,
-					 RXR_BUF_POOL_ALIGNMENT, 0,
-					 rxr_env.recvwin_size, 0);
+					 RXR_BUF_POOL_ALIGNMENT,
+					 rxr_env.recvwin_size,
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_ep.c:1097]: setting max_cnt as rxr_env.recvwin_size seems arbitrary. Just leave it as 0?

File: prov/efa/src/rxr/rxr_ep.c (hunk 1 above)
Comment (rajachan) [reply prov/efa/src/rxr/rxr_ep.c:1104]: The attribute flags are opt-ins, like how we have the one for huge pages. I don't think you will need to modify providers if the default behavior is left at the lazy allocation; but post 1.13.0 sounds good to me.

File: prov/efa/src/rxr/rxr_ep.c [hunk 3]
```diff
@@ -1089,15 +1081,33 @@ int rxr_ep_init(struct rxr_ep *ep)
 
 		if (ret)
 			goto err_free;
+
+		ret = ofi_bufpool_grow(ep->rx_unexp_pkt_pool);
```
Comment (rajachan) [inline prov/efa/src/rxr/rxr_ep.c:1085]: Doxygen comment for rxr_ep_init() needs to be updated to reflect this change.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:1081]: Good catch! fixed in new revision.

File: prov/efa/src/rxr/rxr_ep.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c]: fixed in new revision.

**Review by wckzhang — APPROVED** [review]

**Comment by wzamazon** [comment]
bot:aws:retest

**Pull request merged**